udata_metrics/client.py

```diff
@@ -1,26 +1,104 @@
+        yesterday = yesterday_dt.strftime("%Y-%m-%d")
+        page_type = f"{measurement}".split(".")[0]
+        log.info(f"Running metrics aggregation for {page_type}")
+        id_key = f"{measurement}".split(".")[1]
+        # TODO: Replace with right timezone for logs in the query
+        agg_query = f"""
+                import "date"
+                from(bucket: "vector-bucket")
+                    |> range(start: {yesterday}T00:00:00.000Z, stop: {today}T00:00:00.000Z)
+                    |> filter(fn: (r) => r._measurement == "{measurement}")
+                    |> map(fn: (r) => ({{r with day: date.truncate(t: r._time, unit: 1d)}}))
+                    |> group(columns: ["{id_key}", "day"])
+                    |> count()
+                    |> duplicate(column: "day", as: "_time")
+                    |> map(fn: (r) => ({{r with _field: "{page_type}", _measurement: "count"}}))
+                    |> to(
+                        bucket: "{current_app.config['METRICS_VECTOR_BUCKET']}",
+                        host: "{current_app.config['METRICS_INFLUX_DSN']['url']}",
+                        org: "{current_app.config['METRICS_INFLUX_DSN']['org']}",
+                        token: "{current_app.config['METRICS_INFLUX_DSN']['token']}"
+                    )
+                """
+        self.client.query_api().query(agg_query)
+        self.client.delete_api().delete(
+            start=f"{yesterday}T00:00:00.000Z",
+            stop=f"{today}T00:00:00.000Z",
+            predicate=f"_measurement=\"{measurement}\"",
+            bucket=current_app.config['METRICS_VECTOR_BUCKET'],
+            org=current_app.config['METRICS_INFLUX_DSN']['org']
+        )
 
-    def get_previous_day_measurements(self, collection, tag):
-        query = f'select * from {collection}_views where time > now() - 24h group by {tag};'
-        result = self.client.query(query)
-        return result
+        retrieve_metrics_query = f"""
+            from(bucket: "vector-bucket")
+                |> range(start: 0, stop: {today}T00:00:00.000Z)
+                |> filter(fn: (r) => (r._field == "{page_type}") and (r._measurement == "count"))
+                |> group(columns: ["{id_key}"])
+                |> sum()
+                |> yield(name: "views-{page_type}")
+        """
+        result = self.client.query_api().query(retrieve_metrics_query)
+        for table in result:
+            for record in table:
+                # Currently we ignore resource hits as they are different from resource downloads.
+                if page_type == "resource_hit":
+                    continue
+                elif page_type == "resource":
+                    # Since resource is not a model per se but part of the dataset model, we need to
+                    # treat it separately.
+                    try:
+                        model_id = uuid.UUID(record[id_key])
+                        model_result = get_resource(model_id)
```

Comment:
Sadly metric aggregation takes too much time, probably due to `get_resource` taking a significant time * the large number of resources.

Reply:
After a bit of inquiry, get_resource can take a lot of time if the targeted resource is in a dataset with numerous resources.
We've tried optimizing the get_resource without success. For info:
- having the dataset id, getting the dataset first before getting the precise resource, but it was not efficient enough either
- querying only relevant resource from Mongo `Dataset.objects.filter(resources__id=id).fields(resources={'$elemMatch': {'_id': str(id)}}).first()`, it would not work at validation time, since the entire dataset hasn't been returned.

However, it appeared that the `.save()` command would also take a big amount of time (more than fetching) for those huge datasets.

The other solution is to iterate on dataset x resources objects directly, updating the metric only if found in the table, and saving the dataset once only. An example solution has been pushed here: https://github.com/opendatateam/udata-metrics/commit/d4f644c7d1a4dd84de610be2aefed9d5e72f540e.
The code is ugly, but shows a proposal with an update of resources metrics running under 5mins locally on my computer.